src/protocol/boolean/random_bits_generator.rs

```diff
@@ -4,142 +4,21 @@ use crate::ff::Field;
-        let x = self.entries.remove(&self.read_pointer);
-        self.read_pointer = self.read_pointer.wrapping_add(1);
-        x.expect("unexpected buffer depletion")
-    }
-}
-
-#[derive(Debug)]
-struct State<F, S, C>
-where
-    F: Field,
-    S: SecretSharing<F>,
-    C: Context<F, Share = S>,
-{
-    buffer: RingBuffer<RandomBitsShare<F, S>>,
-    next_index: u32,
-    abort_count: u32,
-    context: C,
-    _marker: PhantomData<F>,
-}
-
-impl<F, S, C> State<F, S, C>
-where
-    F: Field,
-    S: SecretSharing<F>,
-    C: Context<F, Share = S>,
-{
-    const REFILL_THRESHOLD: usize = 16;
-
-    pub fn new(context: C) -> Self {
-        Self {
-            buffer: RingBuffer::new(),
-            next_index: 0,
-            abort_count: 0,
-            context,
-            _marker: PhantomData::default(),
-        }
-    }
-
-    /// Returns a `RandomBitsShare` instance out of the buffer. If the number
-    /// of buffered items fall below a threshold, it'll replenish.
-    pub async fn get(&mut self) -> Result<RandomBitsShare<F, S>, Error> {
-        // If we have enough shares, take one immediately return
-        if self.buffer.len() > Self::REFILL_THRESHOLD {
-            return Ok(self.buffer.take());
-        }
-
-        // call `SolvedBits` to refill the empty slots in the buffer, and returns
-        // one `RandomBitsShare` instance.
-        let available_space = self.buffer.available_space();
-        try_join_all((0..available_space).map(|_| {
-            let c = self.context.clone();
-            let record_id = RecordId::from(self.next_index);
-
-            // Current implementation will cause the context to panic once u32
-            // wraps around. It's probably safe to reuse the same index again
-            // after so many records being processed.
-            self.next_index = self.next_index.wrapping_add(1);
-
-            async move { solved_bits(c, record_id).await }
-        }))
-        .and_then(|new_stock| async move {
-            for x in new_stock {
-                if let Some(r) = x {
-                    self.buffer.insert(r);
-                } else {
-                    // Keep track of how many times `SolvedBits` aborted. This stat
-                    // is for testing and telemetry purposes only.
-                    self.abort_count += 1;
-                }
-            }
-            Ok(self.buffer.take())
-        })
-        .await
-    }
-}
+use std::{
+    marker::PhantomData,
+    sync::atomic::{AtomicU32, AtomicUsize, Ordering},
+};
 
 /// A struct that pre-generates and buffers random sharings of bits from the
 /// `SolvedBits` protocol. Any protocol who wish to use a random-bits can draw
-/// one by calling `take_one()`. It will call `SolvedBits` once the stock falls
-/// below `REFILL_THRESHOLD` until it fills up the empty slots.
+/// one by calling `take_one()`. Making a generator spawns a task that generates
+/// new values asynchronously up to the given capacity.
```

Comment:
The code seems to just generate one `RandomBitsShare` for each `generate()` call?

Reply:
Yeah, I fixed that already.  Failed to push.